src/ShallowWrapper.lua

```diff
@@ -0,0 +1,228 @@
+	end,
+}
+
+local function countChildrenOfElement(element)
+	if ElementKind.of(element) == ElementKind.Fragment then
+		local count = 0
+
+		for _, subElement in pairs(element.elements) do
+			count = count + countChildrenOfElement(subElement)
+		end
+
+		return count
+	else
+		return 1
+	end
+end
+
+local function getChildren(virtualNode, results, maxDepth)
+	if ElementKind.of(virtualNode.currentElement) == ElementKind.Fragment then
+		for _, subVirtualNode in pairs(virtualNode.children) do
+			getChildren(subVirtualNode, results, maxDepth)
+		end
+	else
+		local childWrapper = ShallowWrapper.new(
+			virtualNode,
+			maxDepth
+		)
+
+		table.insert(results, childWrapper)
+	end
+end
+
+local function filterProps(props)
+	if props[Children] == nil then
+		return props
+	end
+
+	local filteredProps = {}
+
+	for key, value in pairs(props) do
+		if key ~= Children then
+			filteredProps[key] = value
+		end
+	end
+
+	return filteredProps
+end
+
+function ShallowWrapper.new(virtualNode, maxDepth)
+	virtualNode = findNextVirtualNode(virtualNode, maxDepth)
+
+	local wrapper = {
+		_virtualNode = virtualNode,
+		_childrenMaxDepth = maxDepth - 1,
+		_virtualNodeChildren = maxDepth == 0 and {} or virtualNode.children,
+		_shallowChildren = nil,
+		type = getTypeFromVirtualNode(virtualNode),
+		props = filterProps(virtualNode.currentElement.props),
+		hostKey = virtualNode.hostKey,
+		instance = virtualNode.hostObject,
+	}
+
+	return setmetatable(wrapper, ShallowWrapperMetatable)
+end
+
+function ShallowWrapper:childrenCount()
+	local count = 0
+
+	for _, virtualNode in pairs(self._virtualNodeChildren) do
+		local element = virtualNode.currentElement
+		count = count + countChildrenOfElement(element)
+	end
+
+	return count
+end
+
+function ShallowWrapper:find(constraints)
+	for constraint in pairs(constraints)	 do
+		if not ContraintFunctions[constraint] then
+			error(('unknown constraint %q'):format(constraint))
+		end
+	end
+
+	local results = {}
+	local children = self:getChildren()
+
+	for i=1, #children do
+		local childWrapper = children[i]
+		if childWrapper:_satisfiesAllContraints(constraints) then
+			table.insert(results, childWrapper)
+		end
+	end
+
+	return results
+end
+
+function ShallowWrapper:findUnique(constraints)
+	local children = self:getChildren()
+
+	if constraints == nil then
+		assert(
+			#children == 1,
+			("expect to contain exactly one child, but found %d"):format(#children)
+		)
+		return children[1]
+	end
+
+	local constrainedChildren = self:find(constraints)
+
+	assert(
+		#constrainedChildren == 1,
+		("expect to find only one child, but found %d"):format(#constrainedChildren)
+	)
+
+	return constrainedChildren[1]
+end
+
+function ShallowWrapper:getChildren()
+	if self._shallowChildren then
+		return self._shallowChildren
+	end
+
+	local results = {}
+
+	for _, childVirtualNode in pairs(self._virtualNodeChildren) do
+		getChildren(childVirtualNode, results, self._childrenMaxDepth)
```

Comment:
instead of using a non member function, this could be a bit simpler:

```
local childWrapper = ShallowWrapper.new(
	childVirtualNode,
	self._childrenMaxDepth
)
childWrapper:getChildren()
table.insert(results, childWrapper)
```

this would need the fragment check as well if we got rid of the other function

Reply:
The thing I don't like with the member function is that I would have to copy back the elements from the `getChildren` call into the result table recursively. Do you think it would be better to just go for clarity at this point? I have not benchmark anything yet so I don't have numbers about what is the performance gain

Reply:
Perf is a good point, but I wonder if that concern goes away if we redefine depth such that we're building more deeply nested tables instead of coalescing the children at various levels of the tree to be the shallow children.